x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -0,0 +1,216 @@
+          }          
+          if(this.ids.add(id) == false) {
+              throw new IllegalArgumentException("[" + NAME + "] duplicate id found in list: "+id);              
+          }
+      }
+      
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public PinnedQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        Collections.addAll(ids, in.readStringArray());
+        organicQuery = in.readNamedWriteable(QueryBuilder.class);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeStringArray(ids.toArray(new String[ids.size()]));
+        out.writeNamedWriteable(organicQuery);
+    }
+
+    /**
+     * @return the organic query set in the constructor
+     */
+    public QueryBuilder organicQuery() {
+        return this.organicQuery;
+    }
+
+    /**
+     * Returns the pinned ids for the query.
+     */
+    public Set<String> ids() {
+        return Collections.unmodifiableSet(this.ids);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        if (organicQuery != null) {
+            builder.field(ORGANIC_QUERY_FIELD.getPreferredName());
+            organicQuery.toXContent(builder, params);
+        }
+        builder.startArray(IDS_FIELD.getPreferredName());
+        for (String value : ids) {
+            builder.value(value);
+        }
+        builder.endArray();
+        printBoostAndQueryName(builder);
+        builder.endObject();
+    }
+    
+    
+    
+    private static final ConstructingObjectParser<PinnedQueryBuilder, XContentParser> PARSER = new ConstructingObjectParser<>(NAME,
+            a -> 
+                {
+                    QueryBuilder organicQuery = (QueryBuilder) a[0];
+                    @SuppressWarnings("unchecked")
+                    List<String> ids = (List<String>) a[1];
+                    return new PinnedQueryBuilder(organicQuery, ids.toArray(new String[0]));
+                }
+             );
+    static {
+        PARSER.declareObject(constructorArg(), (p, c) -> parseInnerQueryBuilder(p), ORGANIC_QUERY_FIELD);        
+        PARSER.declareStringArray(constructorArg(), IDS_FIELD);
+        declareStandardFields(PARSER);
+    }
+
+    public static PinnedQueryBuilder fromXContent(XContentParser parser) {
+        try {
+            return PARSER.apply(parser, null);
+        } catch (IllegalArgumentException e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws IOException {
+        if (ids.size() == 0 && organicQuery == null) {
+            return new MatchNoneQueryBuilder();
```

Comment:
the constructor rejects organicQuery==null so this code is unreachable I think?

Reply:
Should constructor error if ids.length ==0?
Also, what is the contract here for rewrites and nested objects?  Should I call organicQuery.rewrite() and if  it changes return a new PinnedQueryBuilder?

Reply:
> Should constructor error if ids.length ==0?

Good question. I think not failing would help use this query within templates so I'm leaning towards not failing.

> Also, what is the contract here for rewrites and nested objects? Should I call organicQuery.rewrite() and if it changes return a new PinnedQueryBuilder?

Yes indeed!